---
title: "R script for haddad-etal-2025"
author: "Mariss Haddad"
date: '2025-10-22'
output: html_document
---

### RUN ON R 4.4.1

# Import libraries & define functions. 

```{r, include=FALSE}
# Import libraries
library(stringr)
library(ggpubr)
library(dplyr)
library(vegan)
library(pheatmap)
library(ape)
library(scales)
library(reshape2)
library(tidyr)
library(ggplot2)
library(cowplot)
library(car)
library(ComplexHeatmap)
library(circlize)
library(patchwork)
library(ggnewscale)

# Geometric mean function
geometric_mean = function(x) {
  exp(mean(log(x[x>0])))
}

# Function to count zeros in each column
count_zeros = function(column) {
  sum(column == 0)
}
```

# Species_adjusted_readCounts were generated using Kraken2 and Braken. They are adjusted by the tool to estimate the true read count across the full genome.

# Biomass refers to (final - intial) OD measurements output by a biolector XT.

```{r}
# Import files 
metadata = read.csv("metadata.csv", header = TRUE)
metadata$donor = as.factor(metadata$donor)
biomass = read.csv("biomass_diff.csv", header = TRUE)
species_adjusted_readCounts = read.csv("species_abunds.csv", header = TRUE, row.names = 1) # 1 media well lost; 143 cols
```

```{r}
# Keep only rows that match to the metadata
temp = data.frame(t(species_adjusted_readCounts))
temp$sequence_id = rownames(temp)
temp = left_join(metadata, temp, by = "sequence_id")
extracted = subset(temp, select = -c(biolector_id, metabolomics_id, growthcurve_id, gridion_id, treatment, barc, sample_id, dose, condition, donor))
extracted = extracted[complete.cases(extracted$sequence_id),]
rownames(extracted) = extracted$sequence_id
extracted = subset(extracted, select = -c(sequence_id))
species_adjusted_readCounts = data.frame(t(extracted))
```

# Start with all the community level microbiome analysis. 

# Total estimated read count

```{r}
# Media control
media_controls = metadata$sequence_id [which(metadata$treatment == "media")]
media_control_count = species_adjusted_readCounts[,which(colnames(species_adjusted_readCounts) %in% media_controls)]
media_control_count_sum = data.frame(samples = colnames(media_control_count), total_reads = colSums(media_control_count))

ggbarplot(media_control_count_sum, x = "samples", y = "total_reads", xlab = "Sample", ylab = "Number_of_reads", title = "Read counts in media control samples") + theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Species these contaminating reads are from
media_control_count_nonZero = media_control_count[which(rowSums(media_control_count) > 0),]
pheatmap(media_control_count_nonZero)
```

# Number of contaminating reads in media controls is very low compared the treatment samples. Therefore this is not of huge concern. However, should be kept in mind for analysis of metabolomics data. We may want to pick the cleanest media control for our baseline.

# Remove media controls for the remainder of the analysis

```{r}
species_adjusted_readCounts = species_adjusted_readCounts[,which(!colnames(species_adjusted_readCounts) %in% media_controls)]
```

# Read counts by treatment. Identification and removal of outliers.

```{r}
# Histogram across samples
species_adjusted_readCounts_sum = data.frame(sequence_id = colnames(species_adjusted_readCounts), total_reads = colSums(species_adjusted_readCounts))

species_adjusted_readCounts_sum = left_join(species_adjusted_readCounts_sum, metadata)

# Remove NA values
species_adjusted_readCounts_sum = species_adjusted_readCounts_sum[complete.cases(species_adjusted_readCounts_sum),]

ggboxplot(species_adjusted_readCounts_sum, x = "treatment", y = "total_reads", xlab = "Sample", ylab = "Number_of_reads", title = "Read counts per treatment") + theme(axis.text.x = element_text(angle = 45, hjust = 1))

summary(species_adjusted_readCounts_sum$total_reads)
```

# Shannon diversity by treatment. 

```{r}
# Shannon diversity
shannon_diversity = data.frame(diversity(t(species_adjusted_readCounts), index = "shannon"))
colnames(shannon_diversity) = c("shannon_diversity")
shannon_diversity$sequence_id = rownames(shannon_diversity)
shannon_diversity_meta = left_join(metadata, shannon_diversity, by = "sequence_id")

# Remove media wells for analysis
shannon_diversity_meta = shannon_diversity_meta[which(shannon_diversity_meta$treatment != "media"),]

# Plot shannon diversity by treatment
ggboxplot(shannon_diversity_meta, x = "treatment", y = "shannon_diversity", xlab = "Treatment", ylab = "Shannon_diversity", title = "Shannon diversity by treatment") + theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

# PCoA rarefied by Bray-Curtis distance. 

```{r, fig.width=5, fig.height=5}
# Rarefy by smallest number of reads above 4000 and calculate Bray-Curtis dissimilarity
species_adjusted_readCounts_rarefied = species_adjusted_readCounts[,which(colSums(species_adjusted_readCounts) > 4000)]
smallest_read_count = min(colSums(species_adjusted_readCounts_rarefied))
distances = avgdist(t(species_adjusted_readCounts_rarefied), dmethod = "bray", sample = smallest_read_count)

# Give a report on the lost barcode(s)
print(paste(ncol(species_adjusted_readCounts) - ncol(species_adjusted_readCounts_rarefied), "barcode(s) lost to rarefaction.", ncol(species_adjusted_readCounts_rarefied), "remain."))

# One barcode was lost; it is a catalyst_med condition with 373 total reads, which is very low compared to the next smallest read count, 10824. 
lost_barcodes = colnames(species_adjusted_readCounts)[!colnames(species_adjusted_readCounts) %in% colnames(species_adjusted_readCounts_rarefied)]

# Generate the PCoA, format it, and join the metadata to it
pcoa_obj = pcoa(distances)
axis_pcoa = as.data.frame(pcoa_obj$vectors)
axis_pcoa$sequence_id = rownames(axis_pcoa)
axis_pcoa_meta = left_join(axis_pcoa, metadata)

# Check if the data frame contains any NAs
print(paste("There are", sum(is.na(axis_pcoa_meta)), "NAs in the data frame."))
```

# Relative abundances for genus: relabunds_genus

```{r}
# Pull out genus
genus_abunds = species_adjusted_readCounts
genus_abunds$genus = str_split(rownames(species_adjusted_readCounts), pattern = "_", simplify = T)[,1]

# Sum all reads per genus
genus_readsums = genus_abunds %>%
  group_by(genus) %>%
  summarize_all(sum)

# Make genus the row name
genus_readsums = as.data.frame(genus_readsums)
rownames(genus_readsums) = genus_readsums$genus
genus_readsums = genus_readsums[,-1]

# Divide each genus number of reads by total number of reads to calculate relative abundances 
relabunds_genus = sweep(genus_readsums, 2, colSums(genus_readsums), "/")
```

# Relative abundances for species: relabunds_species

```{r}
# Keep species names
species_abunds = species_adjusted_readCounts

# Change cells with reads under 100 to 0
species_abunds[species_abunds < 100] = 0

# Divide each species number of reads by total number of reads to calculate relative abundances
relabunds_species = sweep(species_abunds, 2, colSums(species_abunds), "/")
```

# Absolute abundances for genus: absabunds_genus

```{r}
# Transpose relative abundances
relative_abundances_transposed = data.frame(t(relabunds_genus))
relative_abundances_transposed["sequence_id"] = rownames(relative_abundances_transposed)

# Subset metadata to keep only growthcurve_id and biomass_id
metadata_sub = metadata[c("growthcurve_id", "sequence_id")]

# Merge biomass with metadata
biomass_meta = left_join(metadata_sub, biomass, by = "growthcurve_id")

# Remove growthcurve_id
biomass_meta = subset(biomass_meta, select = -c(growthcurve_id))

# Merge biomass_meta with relative_abundances_transposed
biomass_rel = left_join(biomass_meta, relative_abundances_transposed, by = "sequence_id")

# Drop rows where 3:ncol(biomass_rel) are NA
biomass_rel = biomass_rel[complete.cases(biomass_rel[,3:ncol(biomass_rel)]),]

# For each row, multiply columns 3:ncol(biomass_rel) by column 2
abs_abunds_genus_t = biomass_rel %>% mutate(across(3:ncol(biomass_rel), ~ . * biomass_diff))

# Remove biomass_diff
abs_abunds_genus_t = subset(abs_abunds_genus_t, select = -c(biomass_diff))

# Transpose and format
absabunds_genus_pre = abs_abunds_genus_t
rownames(absabunds_genus_pre) = abs_abunds_genus_t$sequence_id
absabunds_genus_pre = subset(absabunds_genus_pre, select = -c(sequence_id))
absabunds_genus = data.frame(t(absabunds_genus_pre))
```

# Absolute abundances for species: absabunds_species

```{r}
# Transpose relative abundances
relative_abundances_species_t = data.frame(t(relabunds_species))
relative_abundances_species_t["sequence_id"] = rownames(relative_abundances_species_t)

# Merge biomass_meta with transposed species relative abundances
biomass_rel_species = left_join(biomass_meta, relative_abundances_species_t, by = "sequence_id")

# Drop rows where 3:ncol(biomass_rel_species) are NA
biomass_rel_species = biomass_rel_species[complete.cases(biomass_rel_species[,3:ncol(biomass_rel_species)]),]

# For each row, multiply columns 3:ncol(biomass_rel_species
abs_abunds_species_t = biomass_rel_species %>% mutate(across(3:ncol(biomass_rel_species), ~ . * biomass_diff))

# Remove biomass_diff
abs_abunds_species_t = subset(abs_abunds_species_t, select = -c(biomass_diff))

# Transpose and format
absabunds_species_pre = abs_abunds_species_t
rownames(absabunds_species_pre) = abs_abunds_species_t$sequence_id
absabunds_species_pre = subset(absabunds_species_pre, select = -c(sequence_id))
absabunds_species = data.frame(t(absabunds_species_pre))
```

# Import targeted metabolomics. 

```{r}
# Import targeted metabolomics
targeted_metabolomics = read.csv("targeted_metabolomics.csv", header = TRUE)

# Join to metadata
targeted_metabolomics_meta = left_join(metadata, targeted_metabolomics, by = "metabolomics_id")

# Calculate glutamate amount in condition
var = "media"
glut_var = targeted_metabolomics_meta[targeted_metabolomics_meta$treatment==var,]
glut_amt = mean(glut_var$glutamate)
glut_amt

# Remove glutamate
targeted_metabolomics_meta = subset(targeted_metabolomics_meta, select = -c(glutamate))
```

## Now I have relative abundances & absolute abundances by species and genus. I also have targeted metabolomics. 

# Species barplots. 

```{r}
# Subset relative abundances for top 15 most abundant species
top15 = names(sort(rowSums(species_adjusted_readCounts), decreasing = T))[c(1:15)]
top15_rel = species_adjusted_readCounts[top15,]

# Melt data frame
top15_rel_melt = melt(top15_rel)
top15_rel_melt$species = rep(rownames(top15_rel), length(top15_rel[1,]))
colnames(top15_rel_melt)[1] = "sequence_id"

# Join to metadata
top15_rel_melt_merge = left_join(top15_rel_melt, metadata, by = "sequence_id")
top15_rel_melt_merge$unique_id = paste(top15_rel_melt_merge$donor, top15_rel_melt_merge$biolector_id, sep = "_")
```

```{r, fig.width=7, fig.height=7}
for(i in unique(top15_rel_melt_merge$treatment)){
  
    sub_tab = top15_rel_melt_merge[which(top15_rel_melt_merge$treatment == i),]
    
    print(ggplot(sub_tab, aes(fill = species, y = value, x = unique_id)) + 
            geom_bar(position='stack', stat='identity') + 
            scale_color_manual(values = color_palette) + 
            xlab("") + 
            theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = - 0.1, size = 20)) + 
            theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), 
                  axis.line = element_blank(), axis.ticks = element_blank(), legend.title = element_blank(), 
                  text = element_text(size = 20)) + 
            ylab("relative_abundance") + ggtitle(i)
    )
}
```

# Add metadata to absolute genus and species counts: absabunds_genus_meta & absabunds_species_meta

```{r} 
# For each genus, subset absabunds_genus
absabunds_genus_meta = metadata[which(metadata$treatment != "media"),]

for(i in unique(rownames(absabunds_genus))){

  # Locate genus row
  genus_pull = absabunds_genus[which(rownames(absabunds_genus) == i),]
  genus = data.frame(t(genus_pull))
  genus$sequence_id = rownames(genus)
  rownames(genus) = NULL

  # Append dataframe to metadata
  absabunds_genus_meta = left_join(absabunds_genus_meta, genus, by = "sequence_id")
}

# For each species, subset absabunds_species
absabunds_species_meta = metadata[which(metadata$treatment != "media"),]

for (i in unique(rownames(absabunds_species))){

  # Locate species row
  species_pull = absabunds_species[which(rownames(absabunds_species) == i),]
  species = data.frame(t(species_pull))
  species$sequence_id = rownames(species)
  rownames(species) = NULL

  # Append dataframe to metadata
  absabunds_species_meta = left_join(absabunds_species_meta, species, by = "sequence_id")
}
```

# Add metadata to relative genus and species counts: relabunds_genus_meta & relabunds_species_meta

```{r} 
# For each genus, subset absabunds_genus
relabunds_genus_meta = metadata[which(metadata$treatment != "media"),]

for(i in unique(rownames(relabunds_genus))){

  # Locate genus row
  genus_pull = relabunds_genus[which(rownames(relabunds_genus) == i),]
  genus = data.frame(t(genus_pull))
  genus$sequence_id = rownames(genus)
  rownames(genus) = NULL

  # Append dataframe to metadata
  relabunds_genus_meta = left_join(relabunds_genus_meta, genus, by = "sequence_id")
}

# For each species, subset absabunds_species
relabunds_species_meta = metadata[which(metadata$treatment != "media"),]

for (i in unique(rownames(relabunds_species))){

  # Locate species row
  species_pull = relabunds_species[which(rownames(relabunds_species) == i),]
  species = data.frame(t(species_pull))
  species$sequence_id = rownames(species)
  rownames(species) = NULL

  # Append dataframe to metadata
  relabunds_species_meta = left_join(relabunds_species_meta, species, by = "sequence_id")
}
```

# Plot Bifidobacterium genus absolute abundance and gaba by treatment. 

```{r, fig.height=5, fig.width=5}
my_palette_4v = c("#A73030FF","#EFC000FF","#868686FF","#7AA6DCFF")
my_palette_5v = c("#003C67FF","#A73030FF", "#EFC000FF","#868686FF","#7AA6DCFF")

absabunds_genus_meta$condition = factor(absabunds_genus_meta$condition, levels = c("stool", "catalyst", "monomer", "sg-glc"))
targeted_metabolomics_meta$condition = factor(targeted_metabolomics_meta$condition, levels = c("media", "stool", "catalyst", "monomer", "sg-glc"))
```

# Kruskal-Wallis test to get p-value for GABA increase with changing SG-Glc dose. 

```{r}
# Subset for gaba and treatment, for only SG-Glc treatments
kw_test_data = subset(targeted_metabolomics_meta, select = c("gaba", "treatment"))
kw_test_data = kw_test_data[which(grepl("sg-glc", kw_test_data$treatment)),]

# Kruskal-Wallis
kruskal.test(kw_test_data$gaba ~ kw_test_data$treatment)$p.value
```

# Print GABA plot alone. 

```{r, fig.height=6, fig.width=6.5}
plot_1 = ggboxplot(targeted_metabolomics_meta, x = "treatment", y = "gaba",
                    fill = "condition", xlab = "Treatment", ylab = "GABA (mg/mL)",
                    title = "GABA Concentration", palette = my_palette_5v) +
          theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
          theme(plot.title = element_text(hjust = 0.5))

print(plot_1)
```

# Plot Bifidobacterium_adolescentis relative abundance by treatment. 

```{r, fig.height=5, fig.width=4}
# Plot
temp = relabunds_species_meta
temp$Bifidobacterium_adolescentis = temp$Bifidobacterium_adolescentis * 100
temp$condition = factor(temp$condition, levels = c("media", "stool", "catalyst", "monomer", "sg-glc"))

plot_2 = ggboxplot(temp, x = "treatment", y = "Bifidobacterium_adolescentis",
                   fill = "condition", xlab = "Treatment", ylab = "Bifidobacterium adolescentis (%)",
                   palette = my_palette_4v) +
          theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
          theme(plot.title = element_text(hjust = 0.5)) + theme(legend.position = "bottom")

print(plot_2)
```

# Plot 4 Bifidobacterium species relative abundances by treatment. 

```{r}
# Keep rows with "stool" or "sg-glc" in column "condition"
temp2 = relabunds_species_meta[which(grepl("stool", relabunds_species_meta$condition) | grepl("sg-glc", relabunds_species_meta$condition)),]

# Get percentages for rel abunds
temp2$Bifidobacterium_adolescentis = temp2$Bifidobacterium_adolescentis * 100
temp2$Bifidobacterium_breve = temp2$Bifidobacterium_breve * 100
temp2$Bifidobacterium_pseudocatenulatum = temp2$Bifidobacterium_pseudocatenulatum * 100
temp2$Bifidobacterium_bifidum = temp2$Bifidobacterium_bifidum * 100
```

```{r}
# Define custom order
custom_order_temp2 = c("stool", "sg-glc_low", "sg-glc_med", "sg-glc_high")
temp2$treatment = factor(temp2$treatment, levels = custom_order_temp2)

# Reshape the dataframe to long format
temp2_long = temp2 %>%
  pivot_longer(cols = c(Bifidobacterium_adolescentis, Bifidobacterium_breve, Bifidobacterium_pseudocatenulatum, Bifidobacterium_bifidum), names_to = "Variable", values_to = "Value")

# Split out into different dataframes by donor
temp2_long_donor1 = temp2_long[which(temp2_long$donor == "7393"),]
temp2_long_donor2 = temp2_long[which(temp2_long$donor == "7302"),]
temp2_long_donor3 = temp2_long[which(temp2_long$donor == "6391"),]

# Create the boxplot for Donor 6391
ggplot(temp2_long_donor3, aes(x = treatment, y = Value, fill = Variable)) +
  geom_boxplot() +
  geom_point(aes(color = donor)) +
  facet_wrap(~ Variable, scales = "free_y") +  
  theme_minimal() +
  labs(x = "Treatment", y = "Relative Abundance (%)", title = "Donor 6391") +
  theme(legend.position = "none")  # Remove legend (optional)
```

# Merge abundances with metabolomics: abs_genus_holistic, abs_species_holistic, rel_genus_holistic, rel_species_holistic

```{r}
# Merge absabunds_genus_meta with metabolomics
abs_genus_holistic = left_join(absabunds_genus_meta, targeted_metabolomics, by = "metabolomics_id")

# Merge absabunds_species_meta with metabolomics
abs_species_holistic = left_join(absabunds_species_meta, targeted_metabolomics, by = "metabolomics_id")

# Merge relabunds_genus_meta with metabolomics
rel_genus_holistic = left_join(relabunds_genus_meta, targeted_metabolomics, by = "metabolomics_id")

# Merge relabunds_species_meta with metabolomics
rel_species_holistic = left_join(relabunds_species_meta, targeted_metabolomics, by = "metabolomics_id")
```

# Ordination of high and low bifido and gaba producers. 
# Based on genus relative abundances. 

```{r}
# Use rel_genus_holistic

# threshold = Q3 (75th percentile) + 1.5 * Interquartile range (IQR)
threshold = as.numeric(quantile(rel_genus_holistic$Bifidobacterium, 0.75)) + 
  1.5 * IQR(rel_genus_holistic$Bifidobacterium)
threshold

# Create column bifido_cat where if Bifidobacterium is above threshold, it is "high", otherwise "low"
rel_genus_holistic_ord = rel_genus_holistic
rel_genus_holistic_ord$bifido_level = ifelse(rel_genus_holistic_ord$Bifidobacterium > threshold, "high", "low")

# threshold = Q3 (75th percentile) + 1.5 * Interquartile range (IQR)
thresholdg = as.numeric(quantile(rel_genus_holistic_ord$gaba, 0.75)) + 1.5 * IQR(rel_genus_holistic_ord$gaba)

# Create column gaba_cat where if GABA is above threshold, it is "high", otherwise "low"
rel_genus_holistic_ord$gaba_level = ifelse(rel_genus_holistic_ord$gaba > thresholdg, "high", "low")

# Add bifido_cat, gaba_cat, and biolector_id to axis_pcoa_meta
add = subset(rel_genus_holistic_ord, select = c(bifido_level, gaba_level, biolector_id))
axis_pcoa_add = left_join(axis_pcoa_meta, add, by = "biolector_id")
```

```{r, fig.height=5, fig.width=10}
# Plot PCs 3 and 4

# Bifidobacterium level
plot_3 = ggscatter(axis_pcoa_add, x = "Axis.3", y = "Axis.4", 
                   color = "bifido_level", xlab = "Axis 3", ylab = "Axis 4", size = 3,
                   title = "Bifidobacterium Abundance") + 
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  labs(color = "Bifidobacterium Level", 
       x = paste0("PC3 (", percent(pcoa_obj$values$Relative_eig[3]), ")"), 
       y = paste0("PC4 (", percent(pcoa_obj$values$Relative_eig[4]), ")")) +
  stat_ellipse(aes(group = bifido_level, color = bifido_level)) 

# GABA level
plot_4 = ggscatter(axis_pcoa_add, x = "Axis.3", y = "Axis.4", 
                   color = "gaba_level", xlab = "Axis 3", ylab = "Axis 4", size = 3,
                   title = "GABA Production") + 
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  labs(color = "GABA Level", 
       x = paste0("PC3 (", percent(pcoa_obj$values$Relative_eig[3]), ")"), 
       y = paste0("PC4 (", percent(pcoa_obj$values$Relative_eig[4]), ")")) +
  stat_ellipse(aes(group = gaba_level, color = gaba_level)) 

ggarrange(plot_3, plot_4, ncol = 2, legend = "bottom")
```

# Effect of polymerization on microbiome (PCoA). 

```{r}
# Group into categories with medium and high polymerization
axis_pcoa_add_sub = axis_pcoa_add[grepl("low|med|high|stool", axis_pcoa_add$treatment),]

axis_pcoa_add_sub$high_level_treatment = ifelse(axis_pcoa_add_sub$treatment == "sg-glc_low", 
                                                "low dose", axis_pcoa_add_sub$treatment)

axis_pcoa_add_sub$high_level_treatment = ifelse(axis_pcoa_add_sub$treatment == "sg-glc_med", 
                                                "med dose", axis_pcoa_add_sub$high_level_treatment)

axis_pcoa_add_sub$high_level_treatment = ifelse(axis_pcoa_add_sub$treatment == "sg-glc_high", 
                                                "high dose", axis_pcoa_add_sub$high_level_treatment)

axis_pcoa_add_sub$high_level_treatment = ifelse(grepl("catalyst", axis_pcoa_add_sub$high_level_treatment), 
                                                "catalyst", axis_pcoa_add_sub$high_level_treatment)

axis_pcoa_add_sub$high_level_treatment = ifelse(grepl("monomer", axis_pcoa_add_sub$high_level_treatment), 
                                                "monomer", axis_pcoa_add_sub$high_level_treatment)
```

# PCoAs and their boxplots

```{r, fig.height=8, fig.width=6}
plot_5 = ggscatter(axis_pcoa_add_sub, x = "Axis.3", y = "Axis.4", color = "high_level_treatment", palette = "jco",
                   xlab = "Axis 3", ylab = "Axis 4", title = "Polymerization Effect on Microbiome") + 
  stat_ellipse(aes(x = Axis.3, y = Axis.4, group = high_level_treatment, color = high_level_treatment), size = 0.8) + 
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  xlim(-0.3, 0.65) +
  labs(color = "", x = paste0("PC3 (", percent(pcoa_obj$values$Relative_eig[3]), ")"), 
       y = paste0("PC4 (", percent(pcoa_obj$values$Relative_eig[4]), ")"))


custom_order_box = c("catalyst", "stool", "monomer", "low dose", "med dose", "high dose")
axis_pcoa_add_sub$high_level_treatment = factor(axis_pcoa_add_sub$high_level_treatment, levels = custom_order_box)


custom_colors = c(
  "catalyst" = "#f8766d",
  "stool" = "#b79f00",
  "monomer" = "#00ba38",
  "low dose" = "#00bfc4",
  "med dose" = "#649cfb",
  "high dose" = "#f56fda"
)

box_1 = ggplot(axis_pcoa_add_sub, aes(x = Axis.3, y = high_level_treatment, fill = high_level_treatment)) +
  geom_boxplot(alpha = 0.7) +
  #geom_jitter(width = 0, height = 0.2, alpha = 0.5, size = 1) +
  scale_fill_manual(values = custom_colors) +
  labs(x = "PC3 (16%)", y = "") +
  xlim(-0.25, 0.65) +
  theme_minimal() +
  theme(
    legend.position = "none", 
    plot.title = element_text(hjust = 0.5),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line.x = element_line(color = "black"),
    axis.line.y = element_line(color = "black")
  )

# Marginal plots in ggpubr
combo_1 = plot_5 / box_1 + plot_layout(heights = c(3, 1))
combo_1
```


# Separate PCoA colored by GABA with threshold in the legend. 

```{r, fig.height=8, fig.width=6}
axis_pcoa_add_legend = axis_pcoa_add
axis_pcoa_add_legend$gaba_level = ifelse(axis_pcoa_add_legend$gaba == "low", "low (\u2264 0.28 mg/mL)", "high (>0.28 mg/mL)")

plot_6 = ggscatter(axis_pcoa_add_legend, x = "Axis.3", y = "Axis.4", 
                   color = "gaba_level", xlab = "Axis 3", ylab = "Axis 4", title = "GABA Production") + 
  stat_ellipse(aes(x = Axis.3, y = Axis.4, group = gaba_level, color = gaba_level), size = 0.8) + 
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  xlim(-0.3, 0.65) +
  scale_color_manual(values = c("tan2", "violet")) + 
  labs(color = "GABA Level", 
       x = paste0("PC3 (", percent(pcoa_obj$values$Relative_eig[3]), ")"), 
       y = paste0("PC4 (", percent(pcoa_obj$values$Relative_eig[4]), ")"))

custom_colors = c(
  "low (≤ 0.28 mg/mL)" = "violet",
  "high (>0.28 mg/mL)" = "tan2"
)

box_2 = ggplot(axis_pcoa_add_legend, aes(x = Axis.3, y = gaba_level, fill = gaba_level)) +
  geom_boxplot(alpha = 0.7) +
  #geom_jitter(width = 0, height = 0.2, alpha = 0.5, size = 1) +
  scale_fill_manual(values = custom_colors) +
  labs(x = "PC3 (16%)", y = "") +
  xlim(-0.3, 0.65) +
  theme_minimal() +
  theme(
    legend.position = "none", 
    plot.title = element_text(hjust = 0.5),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line.x = element_line(color = "black"),
    axis.line.y = element_line(color = "black")
  )

combo_2 = plot_6 / box_2 + plot_layout(heights = c(3, 1))
combo_2
```

# Separate PCoA showing donor effect on PC1 and PC2

```{r}
# Change axis_pcoa_add_legend column donors from numbers to A B C
axis_pcoa_add_legend$donor = ifelse(axis_pcoa_add_legend$donor == "6391", "A", 
                                    ifelse(axis_pcoa_add_legend$donor == "7302", "B", 
                                           ifelse(axis_pcoa_add_legend$donor == "7393", "C",
                                                  axis_pcoa_add_legend$donor)))

# Redefining color scheme for robustness
point_colors = c(
  "low (≤ 0.28 mg/mL)" = "violet",
  "high (>0.28 mg/mL)" = "tan2"
)

ellipse_colors = c(
  "A" = "#f8766d",
  "B" = "#00ba38",
  "C" = "#649cfb"
)

plot_7 = ggscatter(axis_pcoa_add_legend, x = "Axis.1", y = "Axis.2", 
                   color = "donor",
                   xlab = "Axis 1", ylab = "Axis 2", title = "Donor Signal") + 
  scale_color_manual(values = ellipse_colors, name = "Donor") + 
  stat_ellipse(aes(x = Axis.1, y = Axis.2, group = donor, color = donor), size = 0.8) + 
  theme(axis.text.x = element_text(angle = 0, hjust = 1)) + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  xlim(-0.55, 0.55) +
  labs(x = paste0("PC1 (", percent(pcoa_obj$values$Relative_eig[1]), ")"), 
       y = paste0("PC2 (", percent(pcoa_obj$values$Relative_eig[2]), ")"))

#plot(plot_7)
```

```{r, fig.height=8, fig.width=6}
box_3 = ggplot(axis_pcoa_add_legend, aes(x = Axis.1, y = donor, fill = donor)) +
  geom_boxplot(alpha = 0.7) +
  #geom_jitter(width = 0, height = 0.2, alpha = 0.5, size = 1) +
  scale_fill_manual(values = ellipse_colors) +
  labs(x = "PC1 (46%)", y = "") +
  xlim(-0.55, 0.55) +
  theme_minimal() +
  theme(
    legend.position = "none", 
    plot.title = element_text(hjust = 0.5),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.line.x = element_line(color = "black"),
    axis.line.y = element_line(color = "black")
  )
#box_3
combo_3 = plot_7 / box_3 + plot_layout(heights = c(3, 1))
combo_3
```
















# Find top 20 species based on positive correlation with GABA across all donors and sg-glc treatments: top20_sp
# Using positive correlation because I am looking for GABA producers

```{r}
# Subset absabunds_species_meta and change donors to A B C
abs_species = subset(absabunds_species_meta, select = -c(sequence_id, metabolomics_id, growthcurve_id, gridion_id, barc, sample_id, dose, condition))
abs_species$donor = ifelse(abs_species$donor == "6391", "A", ifelse(abs_species$donor == "7302", "B", "C"))

# Add in gaba values
gaba_vals = subset(abs_species_holistic, select = c(biolector_id, gaba))
abs_species_gaba = left_join(abs_species, gaba_vals, by = "biolector_id")
abs_species_gaba = subset(abs_species_gaba, select = -c(biolector_id))

# Get correlation between each species and gaba 
sp = colnames(abs_species_gaba)[3:(ncol(abs_species_gaba) - 1)]
correlation = data.frame()

for (i in sp){
  
  # Subset for columns gaba and species i
  sub = abs_species_gaba[, c("gaba", i)]
  
  # If the species has no reads, skip it
  if (sum(sub[,2]) == 0) {
    next
  }
  
  # Calculate correlations and save to dataframe
  cor = cor.test(sub[,1], sub[,2])$estimate
  correlation = rbind(correlation, data.frame(species = i, cor = cor))
}

# Sort correlation dataframe by positive cor (to find GABA producers)
correlation = correlation[order(-correlation$cor),]

# Remove B. ruminantium because it is a misread of adolescentis
correlation = correlation[which(correlation$species != "Bifidobacterium_ruminantium"),]

# Save top 20 species with correlation to gaba
top20_sp = correlation[1:20,] 
```

## Define functions to handle data for heatmap creation. 

```{r}
# Function to group data by treatment
group_and_summarize = function(data) {

  # Remove donor
  data = subset(data, select = -c(donor))

  # Group by column "treatment" using average
  data = data %>%
    group_by(treatment) %>%
    summarise_all(mean)

  # Make "treatment" the row name and remove column "treatment"
  data = data.frame(data)
  rownames(data) = data$treatment
  data = subset(data, select = -c(treatment))

  return(data)
}
```

```{r}
# Function to get and format all species abundances
get_all_abunds = function(data, abunds_type = "rel_abunds") {
  all_abunds = list()
  
  for (X in c("sg-glc_low", "sg-glc_med", "sg-glc_high")) {
  
    # Subset and transpose data, add numeric column Y
    abunds_X = data[rownames(data) == X, , drop = FALSE]
    abunds_X_t = data.frame(t(abunds_X))
    abunds_X_t$Y = as.numeric(abunds_X_t[, 1])
  
    # Sort in descending order and remove column Y
    abunds_X_t = abunds_X_t[order(-abunds_X_t$Y), , drop = FALSE]
    abunds_X_t = subset(abunds_X_t, select = -Y)
    
    # Make rownames into column species, remove rownames
    abunds_X_t$species = rownames(abunds_X_t)
    rownames(abunds_X_t) = NULL
  
    # Save in a unique dataframe
    if (abunds_type == "rel_abunds") {
      all_abunds[[paste0("relabunds_", X)]] = abunds_X_t
    } else {
      all_abunds[[paste0("absabunds_", X)]] = abunds_X_t
    }
  }
    
  return(all_abunds)
}
```

```{r}
# Function for final formatting of tables
final_sort = function(data, abunds_type = "rel_abunds") {
  
  if (abunds_type == "rel_abunds") {
    # Rename column to rel_abunds
    colnames(data)[3] = "rel_abunds"
  
    # Sort by columns "cor" and "rel_abunds"
    data = data[order(-data$cor, -data$rel_abunds),]
    
  } else {
    
    # Rename column to abs_abunds
    colnames(data)[3] = "abs_abunds"
  
    # Sort by columns "cor" and "abs_abunds"
    data = data[order(-data$cor, -data$abs_abunds),]
  }
}
```

```{r, fig.width=4, fig.height=6}
## Function to create and draw heatmaps for a given table

create_heatmaps = function(data_table, abunds_type = "rel_abunds", global_min_abunds, global_max_abunds, global_min_cor, global_max_cor, col_title) {
  
  # Prepare the data
  rownames(data_table) = data_table$species
  data_table = subset(data_table, select = -c(species))
  
  # Get abundances column
  if (abunds_type == "rel_abunds") {
    abunds_col = as.matrix(subset(data_table, select = c("rel_abunds"), drop = FALSE))
  } else {
    abunds_col = as.matrix(subset(data_table, select = c("abs_abunds"), drop = FALSE))
  }

  # Get correlation column
  cor_col = as.matrix(subset(data_table, select = c("cor"), drop = FALSE))

  # Define color scales via global max and min cor and abunds
  cor_col_fun = colorRamp2(c(global_min_cor, global_max_cor), c("white", "#A73030FF")) 
  abunds_col_fun = colorRamp2(c(global_min_abunds, global_max_abunds), c("white", "#7AA6DCFF")) 
  
  # Create heatmaps
  ht_cor = Heatmap(cor_col, col = cor_col_fun, name = "cor", column_title = "Correlation", 
                   row_names_gp = gpar(fontsize = 8), row_names_side = "left",
                   column_title_gp = gpar(fontsize = 10), row_title_gp = gpar(fontsize = 8),
                   cluster_rows = FALSE, show_row_names = TRUE, show_column_names = FALSE,
                   rect_gp = gpar(col = "gray30", lwd = 0.5), column_gap = unit(0, "mm"))
  
  abunds_name = ifelse(abunds_type == "rel_abunds", "rel_abund", "abs_abund")

  ht_abunds = Heatmap(abunds_col, col = abunds_col_fun, name = abunds_name, 
                      column_title = col_title, 
                      row_names_gp = gpar(fontsize = 8), row_names_side = "left", 
                      column_title_gp = gpar(fontsize = 10), 
                      row_title_gp = gpar(fontsize = 8), cluster_rows = FALSE, 
                      show_row_names = TRUE, show_column_names = FALSE,
                      rect_gp = gpar(col = "gray30", lwd = 0.5), column_gap = unit(0, "mm"))

  return(list(ht_cor = ht_cor, ht_abunds = ht_abunds))
}
```

```{r, fig.width=4, fig.height=6}
# Function to draw heatmaps 
draw_heatmaps = function(table_A, table_B, table_C, abunds_type = "rel_abunds") {
  
  # Get global max and min correlation and abundance
  all_tables = list(get(table_A), get(table_B), get(table_C))
  global_min_abunds = min(sapply(all_tables, function(x) min(x[[abunds_type]])))
  global_max_abunds = max(sapply(all_tables, function(x) max(x[[abunds_type]])))
  global_min_cor = min(sapply(all_tables, function(x) min(x$cor)))
  global_max_cor = max(sapply(all_tables, function(x) max(x$cor)))
  
  # Create heatmaps for each table
  heatmaps_A = create_heatmaps(get(table_A), abunds_type, global_min_abunds, global_max_abunds, global_min_cor, global_max_cor, "Low")
  heatmaps_B = create_heatmaps(get(table_B), abunds_type, global_min_abunds, global_max_abunds, global_min_cor, global_max_cor, "Med")
  heatmaps_C = create_heatmaps(get(table_C), abunds_type, global_min_abunds, global_max_abunds, global_min_cor, global_max_cor, "High")
  
  # Combine heatmaps for each table side by side
  ht_list_A = heatmaps_A$ht_cor + heatmaps_A$ht_abunds
  ht_list_B = heatmaps_B$ht_abunds
  ht_list_C = heatmaps_C$ht_abunds
  
  # Combine the three heatmaps into a grid layout
  combined_ht = ht_list_A + ht_list_B + ht_list_C
  
  # Draw the combined heatmaps on a single plot
  draw(combined_ht, heatmap_legend_side = "right", annotation_legend_side = "right", ht_gap = unit(0, "mm"))
}
```

## Correlation & Abundances Heatmaps

# Make heatmaps from the correlation and absolute abundance data for the 20 selected species by sg-glc dose. 

```{r}
# Subset absabunds_species_meta and change donors to A B C
abs_species = subset(absabunds_species_meta, select = -c(biolector_id, sequence_id, metabolomics_id, growthcurve_id, gridion_id, barc, sample_id, dose, 
                                                         condition))
abs_species$donor = ifelse(abs_species$donor == "6391", "A", ifelse(abs_species$donor == "7302", "B", "C"))

# Group and summarize data
abs_species_treatment = group_and_summarize(abs_species)

# Get all species absabunds by dose
abs_species_treatment_low = get_all_abunds(abs_species_treatment, "abs_abunds")$"absabunds_sg-glc_low"
abs_species_treatment_med = get_all_abunds(abs_species_treatment, "abs_abunds")$"absabunds_sg-glc_med"
abs_species_treatment_high = get_all_abunds(abs_species_treatment, "abs_abunds")$"absabunds_sg-glc_high"

# Join correlation tables with abs abunds tables
table_sgglc_low_r = left_join(top20_sp, abs_species_treatment_low, by = "species")
table_sgglc_med_r = left_join(top20_sp, abs_species_treatment_med, by = "species")
table_sgglc_high_r = left_join(top20_sp, abs_species_treatment_high, by = "species")

# Name and sort columns "cor" and "abs_abunds"
table_sgglc_low_abs = final_sort(table_sgglc_low_r, "abs_abunds")
table_sgglc_med_abs = final_sort(table_sgglc_med_r, "abs_abunds")
table_sgglc_high_abs = final_sort(table_sgglc_high_r, "abs_abunds")
```

```{r, fig.width=6, fig.height=8}
# Draw heatmaps
draw_heatmaps("table_sgglc_low_abs", "table_sgglc_med_abs", "table_sgglc_high_abs", "abs_abunds")
```


# ***************************************

# Investigate bifido relationship with gaba. 

```{r}
cor(abs_genus_holistic$gaba, abs_genus_holistic$Bifidobacterium)

# Plot gaba vs Bifidobacterium with best fit line
ggscatter(abs_genus_holistic, x = "gaba", y = "Bifidobacterium", xlab = "GABA", ylab = "Bifidobacterium", title = "GABA vs Bifidobacterium") + geom_smooth(method = "lm")

# Make a list of all species (i.e. column names) containing "Bifidobacterium"
bifido_list = colnames(absabunds_species_meta)[grep("Bifidobacterium", colnames(absabunds_species_meta))]
```

# Subset for sg-glc gaba producers. 

```{r}
# Subset for sg-glc only
species_sgglc = abs_species_holistic[which(abs_species_holistic$condition == "sg-glc"),]

# Look at gaba distribution
species_sgglc$gaba
ggplot(species_sgglc, aes(x = gaba)) + geom_histogram()

# Remove columns 1-12
species_sgglc_subset = subset(species_sgglc, select = -c(1:12))

# Subset species_sg-glc_subset to include columns containing "gaba", "bacteroides", "bifidobacterium", "lactobacillus", and "streptococcus"
species_sgglc_subset = species_sgglc_subset[,c(grep("gaba", colnames(species_sgglc_subset)), grep("Bifidobacterium", colnames(species_sgglc_subset)), grep("Lactobacillus", colnames(species_sgglc_subset)), grep("Streptococcus", colnames(species_sgglc_subset)), grep("Bacteroides", colnames(species_sgglc_subset)))]
species_sgglc_subset = subset(species_sgglc_subset, select = -c(Evtepia_gabavorous))

# Remove all columns containing only zeros
species_sgglc_subset = species_sgglc_subset[, colSums(species_sgglc_subset != 0) > 0]

# Calculate correlation coefficient between gaba and each species
correlations = data.frame(sapply(species_sgglc_subset, function(x) cor(species_sgglc_subset$gaba, x)))
colnames(correlations)[1] = "corr"

# Sort correlations in descending order without changing type
correlations = correlations[order(-correlations$corr), , drop = FALSE]

# Keep only rows with rownames containing "Bifido"
correlations_bif = correlations[grep("Bifidobacterium", rownames(correlations)), , drop = FALSE]

# Plot Bifidobacterium_ruminantium against Bifidobacterium_adolescentis
ggscatter(species_sgglc_subset, x = "Bifidobacterium_adolescentis", y = "Bifidobacterium_ruminantium") + geom_smooth(method = "lm")
```

# Remove columns with n or more zeros. 

```{r, fig.height=5, fig.width=5}
# Identify columns with 10 or more zeros
columns_to_remove = sapply(species_sgglc_subset, count_zeros) >= 10

# Remove those columns
species_sgglc_cleaned = species_sgglc_subset[, !columns_to_remove]

# Linear model (max 17 predictors)
lm_17 = lm(gaba ~ ., data = species_sgglc_cleaned)
summary(lm_17)

# Calculate VIF to check for multicolinearity 
vif(lm_17) # Two species come up together 

# Calculate colsums for Bifidobacterium_adolescentis and Bifidobacterium_ruminantium
colSums(species_sgglc_cleaned[,c("Bifidobacterium_adolescentis", "Bifidobacterium_ruminantium")])

# Remove the one with lower abundance
species_sgglc_cleaned = subset(species_sgglc_cleaned, select = -c(Bifidobacterium_ruminantium))

# Fit lm with gaba vs bif adolescentis
lm_bifadol = lm(gaba ~ Bifidobacterium_adolescentis, data = species_sgglc_cleaned)
summary(lm_bifadol)
r_sq = round(summary(lm_bifadol)$r.squared,2)

# Plot Bifidobacterium_adolescentis vs GABA concentration
ggscatter(species_sgglc_cleaned, x = "gaba", y = "Bifidobacterium_adolescentis", 
          xlab = "GABA (mg/mL)", ylab = "B. adolescentis biomass") + 
  theme(plot.title = element_text(hjust = 0.5)) + 
  geom_smooth(method = "lm", color = "#A73030FF") +
  annotate("text", x = 0.726, y = 400, label = paste("R² =", r_sq), size = 5, color = "#A73030FF") +
  annotate("text", x = 0.66, y = 360, label = paste("p-value =", 0.0027), size = 5, color = "#A73030FF")
```
